src/utils/PaymentsConverter.ts

```diff
@@ -0,0 +1,236 @@
+export function getPaymentsActivationResponse(
+  iNodoAttivaRPTOutput: InodoAttivaRPTOutput
+): Either<Error, PaymentsActivationResponse> {
+  if (
+    iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.esito === PPTPortTypes.Esito.KO
+  ) {
+    return new Left(new Error(ControllerError.REQUEST_REJECTED));
+  }
+  const errorOrPaymentsActivationResponse = PaymentsActivationResponse.decode({
+    importoSingoloVersamento:
+      iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+        .importoSingoloVersamento,
+    ibanAccredito:
+      iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA.ibanAccredito,
+    causaleVersamento:
+      iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+        .causaleVersamento,
+    enteBeneficiario: {
+      identificativoUnivocoBeneficiario:
+        iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+          .enteBeneficiario.identificativoUnivocoBeneficiario,
+      denominazioneBeneficiario:
+        iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+          .enteBeneficiario.denominazioneBeneficiario,
+      codiceUnitOperBeneficiario:
+        iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+          .enteBeneficiario.codiceUnitOperBeneficiario,
+      denomUnitOperBeneficiario:
+        iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+          .enteBeneficiario.denomUnitOperBeneficiario,
+      indirizzoBeneficiario:
+        iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+          .enteBeneficiario.indirizzoBeneficiario,
+      civicoBeneficiario:
+        iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+          .enteBeneficiario.civicoBeneficiario,
+      capBeneficiario:
+        iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+          .enteBeneficiario.capBeneficiario,
+      localitaBeneficiario:
+        iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+          .enteBeneficiario.localitaBeneficiario,
+      provinciaBeneficiario:
+        iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+          .enteBeneficiario.provinciaBeneficiario,
+      nazioneBeneficiario:
+        iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+          .enteBeneficiario.nazioneBeneficiario
+    },
+    spezzoniCausaleVersamento:
+      iNodoAttivaRPTOutput.nodoAttivaRPTRisposta.datiPagamentoPA
+        .spezzoniCausaleVersamento
+  });
+
+  if (errorOrPaymentsActivationResponse.isLeft()) {
+    return new Left(new Error(ControllerError.ERROR_INVALID_INPUT));
+  }
+  return new Right(errorOrPaymentsActivationResponse.value);
+}
+
+// Convert PaymentsStatusUpdateRequestPagoPa (PagoPa API) to PaymentsStatusUpdateRequest (controller)
+export function getPaymentsStatusUpdateRequest(
+  paymentsStatusUpdateRequestPagoPa: IcdInfoWispInput
+): Either<Error, PaymentsStatusUpdateRequest> {
+  const errorOrPaymentsStatusUpdateRequest = PaymentsStatusUpdateRequest.decode(
```

Comment:
this function is just:

```ts
return PaymentsStatusUpdateRequest.decode(...)
   .getOrElse(new Error(ControllerError.ERROR_INVALID_INPUT)); 
```

Reply:
btw, I think that all these functions which are just wrappers around `decode().getOrElse()` are not useful by themselves, the caller can just use `getOrElse()` and you can get rid of these methods

Reply:
I agree about mapLeft usage, not about function discarding.
PaymentsConverter is the data mapping owner and I think it's very useful to keep separate it as module. It will improve maintainability.

Reply:
right but *actually* it's just decode + mapLeft so if you add a method *now* it's just unnecessary added complexity atm). when (and if) we'll need more logic, we'll refator this one into it's own module

Reply:
I didn't got your point.
PagoPaAPI is using generic types (string\number) but decode() is providing the data mapping concept and it'is validating that generic types.. now.
In the future, decode will be more precise but it's already doing his job there.

About usage: mapLeft will use a function, like:
decode(...).mapLeft( () => { return new Error(ControllerError.ERROR_INVALID_INPUT) } ); 

About ControllerError usage, It's true, there it's not useful.
We could change it just to new Error() because error content will be ignored by controller (It will just check if validation failed or not)

Reply:
if we move the error handling into the calling function (where it should be, considering that decode *already* returns an error in case of failures), this function will be just a wrapper for decode:

https://github.com/teamdigitale/italia-pagopa-proxy/blob/6173ae80c5b7c073747003c8982c73c301ef0bc1/src/utils/PaymentsConverter.ts#L205

```ts
export function getPaymentsStatusUpdateRequest(
  paymentsStatusUpdateRequestPagoPa: IcdInfoWispInput
): Either<Error, PaymentsStatusUpdateRequest> {
return PaymentsStatusUpdateRequest.decode(
    {
      codiceContestoPagamento:
        paymentsStatusUpdateRequestPagoPa.codiceContestoPagamento,
      idPagamento: paymentsStatusUpdateRequestPagoPa.idPagamento
    }
  );
}
```

= not needed

Reply:
if you want to encapsulate the acto of converting between the object it's better to have a conversion method:

```ts
toPaymentsStatusUpdateRequest(paymentsStatusUpdateRequestPagoPa: IcdInfoWispInput) { ... }
```

Reply:
to summarize: it's ok if you want to keep the conversion between types, but let the caller handle the error (just return it)